File: tests/DesktopGrass.Win2D.Tests/SimTests/OceanTests.cs
// OceanTests.cs - Ocean scene (coral blade variant + bubble emitter +
// fish swimmers). Smoke tests that match the conventions used by
// PineTests / AutumnTests.

using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class OceanTests
{
    private const double Monitor1920 = 1920.0;

    private static Sim BuildSim() => BuildSim(Monitor1920);

    private static Sim BuildSim(double width)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, width, Constants.DEFAULT_DENSITY),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(Constants.CANONICAL_TEST_SEED, width);
        sim.ResetEntities(Constants.CANONICAL_TEST_SEED);
        return sim;
    }

    [Fact]
    public void SetSceneOcean_GeneratesAtLeastOneCoral_AndKeepsValuesInRange()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Ocean);

        var corals = sim.Blades.Where(b => b.IsCoral).ToArray();
        Assert.NotEmpty(corals);

        foreach (var b in corals)
        {
            Assert.False(b.IsPine);
            Assert.False(b.IsCactus);
            Assert.False(b.IsMaple);
            Assert.False(b.IsFlower);
            Assert.False(b.IsMushroom);
            Assert.InRange(b.CoralHeight, Constants.CORAL_HEIGHT_MIN, Constants.CORAL_HEIGHT_MAX);
            Assert.InRange(b.CoralWidth,  Constants.CORAL_WIDTH_MIN,  Constants.CORAL_WIDTH_MAX);
            Assert.InRange((int)b.CoralType,     0, Constants.CORAL_TYPE_COUNT  - 1);
            Assert.InRange((int)b.CoralColorIdx, 0, Constants.CORAL_COLOR_COUNT - 1);
        }
    }

    [Fact]
    public void SetSceneOcean_SpawnsInitialFish_AtOrAboveTargetMinimum()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Ocean);

        int fishCount = sim.Entities.Count(e => e.Kind == EntityKind.Fish);
        Assert.True(fishCount >= Constants.FISH_COUNT_MIN, $"Expected at least {Constants.FISH_COUNT_MIN} fish, got {fishCount}");
        Assert.True(fishCount <= Constants.FISH_COUNT_MAX, $"Expected at most {Constants.FISH_COUNT_MAX} fish, got {fishCount}");
    }

    [Theory]
    [InlineData(1920.0, 2)] // scaled 2.5 -> round-half-to-even -> 2
    [InlineData(3456.0, 4)] // scaled 4.5 -> round-half-to-even -> 4
    public void SetSceneOcean_FishCountRoundsHalfToEven(double width, int expected)
    {
        var sim = BuildSim(width);
        sim.SetScene(Scene.Ocean);

        int fishCount = sim.Entities.Count(e => e.Kind == EntityKind.Fish);
        Assert.Equal(expected, fishCount);
    }

    [Fact]
    public void TickEntities_EmitsBubbles_OverTime()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Ocean);

        // Advance ~10 seconds of sim time.
        const double dt = 1.0 / 60.0;
        for (int i = 0; i < 600; i++)
        {
            sim.GlobalTime += dt;
            sim.TickEntities(dt);
        }

        int bubbleCount = sim.Entities.Count(e => e.Kind == EntityKind.Bubble);
        Assert.True(bubbleCount > 0, "Expected at least one bubble after 10s of simulated Ocean ticks.");
    }

    [Fact]
    public void SwitchingFromOceanToGrass_WipesBubblesAndFish()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Ocean);

        const double dt = 1.0 / 60.0;
        for (int i = 0; i < 120; i++)
        {
            sim.GlobalTime += dt;
            sim.TickEntities(dt);
        }
        Assert.Contains(sim.Entities, e => e.Kind == EntityKind.Fish);

        sim.SetScene(Scene.Grass);

        Assert.DoesNotContain(sim.Entities, e => e.Kind == EntityKind.Bubble);
        Assert.DoesNotContain(sim.Entities, e => e.Kind == EntityKind.Fish);
        Assert.DoesNotContain(sim.Blades, b => b.IsCoral);
    }
}

File: tests/DesktopGrass.Native.Tests/src/ocean_tests.cpp
// ocean_tests.cpp
//
// Ocean scene tests (architecture.md §17). Mirror of the Win2D OceanTests so
// the coral blade variant, bubble emitter, and fish swimmers stay in lockstep
// across impls.

#include "TestHelpers.h"
#include "Sim.h"

#include <algorithm>

using namespace desktopgrass;

namespace {

constexpr double kMonitor1920 = 1920.0;

Sim make_ocean_sim(uint64_t seed = CANONICAL_TEST_SEED,
                   double width = kMonitor1920,
                   double density = DEFAULT_DENSITY) {
    Sim sim = sim_init(seed, width, density);
    sim_set_scene(sim, Scene::Ocean);
    return sim;
}

int count_kind(const Sim& sim, EntityKind kind) {
    return static_cast<int>(std::count_if(sim.entities.begin(), sim.entities.end(),
        [kind](const Entity& e) { return e.kind == kind; }));
}

} // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(OceanTests)
{
public:
TEST_METHOD(OceanSceneGeneratesAtLeastOneCoralAndKeepsValuesInRange) {
    Sim sim = make_ocean_sim();

    int coralCount = 0;
    for (const Blade& b : sim.blades) {
        if (!b.isCoral) continue;
        ++coralCount;
        Assert::IsFalse(b.isPine);
        Assert::IsFalse(b.isCactus);
        Assert::IsFalse(b.isMaple);
        Assert::IsFalse(b.isFlower);
        Assert::IsFalse(b.isMushroom);
        Assert::IsTrue(b.coralHeight >= CORAL_HEIGHT_MIN);
        Assert::IsTrue(b.coralHeight <= CORAL_HEIGHT_MAX);
        Assert::IsTrue(b.coralWidth  >= CORAL_WIDTH_MIN);
        Assert::IsTrue(b.coralWidth  <= CORAL_WIDTH_MAX);
        Assert::IsTrue(static_cast<int>(b.coralType)     >= 0);
        Assert::IsTrue(static_cast<int>(b.coralType)     <= CORAL_TYPE_COUNT  - 1);
        Assert::IsTrue(static_cast<int>(b.coralColorIdx) >= 0);
        Assert::IsTrue(static_cast<int>(b.coralColorIdx) <= CORAL_COLOR_COUNT - 1);
    }
    Assert::IsTrue(coralCount > 0);
}

TEST_METHOD(OceanSceneSpawnsInitialFishAtOrAboveTheTargetMinimum) {
    Sim sim = make_ocean_sim();

    const int fishCount = count_kind(sim, EntityKind::Fish);
    Assert::IsTrue(fishCount >= FISH_COUNT_MIN);
    Assert::IsTrue(fishCount <= FISH_COUNT_MAX);
}

TEST_METHOD(OceanFishCountRoundsHalfToEvenDeterministically) {
    // scaled = 2.5 * width / 1920. Widths chosen so scaled lands exactly on a
    // .5 tie; round-half-to-even must pick the even neighbor (NOT half-up),
    // matching C# Math.Round and independent of the FPU rounding mode.
    Sim tie25 = make_ocean_sim(CANONICAL_TEST_SEED, 1920.0); // scaled 2.5 -> 2
    Assert::IsTrue(count_kind(tie25, EntityKind::Fish) == 2);

    Sim tie45 = make_ocean_sim(CANONICAL_TEST_SEED, 3456.0); // scaled 4.5 -> 4
    Assert::IsTrue(count_kind(tie45, EntityKind::Fish) == 4);
}

TEST_METHOD(OceanTickEmitsBubblesOverTime) {
    Sim sim = make_ocean_sim();

    const double dt = 1.0 / 60.0;
    for (int i = 0; i < 600; ++i) {
        sim.globalTime += dt;
        sim_tick_entities(sim, dt);
    }

    Assert::IsTrue(count_kind(sim, EntityKind::Bubble) > 0);
}

TEST_METHOD(SwitchingFromOceanToGrassWipesBubblesAndFish) {
    Sim sim = make_ocean_sim();

    const double dt = 1.0 / 60.0;
    for (int i = 0; i < 120; ++i) {
        sim.globalTime += dt;
        sim_tick_entities(sim, dt);
    }
    Assert::IsTrue(count_kind(sim, EntityKind::Fish) > 0);

    sim_set_scene(sim, Scene::Grass);

    Assert::IsTrue(count_kind(sim, EntityKind::Bubble) == 0);
    Assert::IsTrue(count_kind(sim, EntityKind::Fish) == 0);
    Assert::IsTrue(std::none_of(sim.blades.begin(), sim.blades.end(),
        [](const Blade& b) { return b.isCoral; }));
}

TEST_METHOD(OceanPaletteIsPinnedInScenePalettes) {
    for (int i = 0; i < PALETTE_SIZE; ++i) {
        Assert::IsTrue(SCENE_PALETTES[static_cast<int>(Scene::Ocean)][i] == OCEAN_PALETTE[i]);
    }
}
};
}
